### PR TITLE
Accept JSON config file

### PR DESCRIPTION
**Pull request by live1206**
While docfx.yml and docfx.json both exist, prefer docfx.yml.

### CODE REVIEW

File: src/docfx/config/Config.cs [hunk 1]
```diff
@@ -217,7 +231,16 @@ string GetBranch()
 
         private static (List<Error>, JObject) LoadConfigObject(string filePath)
         {
-            var (errors, config) = YamlUtility.Deserialize<JObject>(File.ReadAllText(filePath));
+            var errors = new List<Error>();
+            JObject config = null;
+            if (filePath.Contains(".yml", StringComparison.OrdinalIgnoreCase))
```
Comment (yufeih) [inline src/docfx/config/Config.cs]: EndsWith

File: src/docfx/config/Config.cs [hunk 2]
```diff
@@ -217,7 +231,16 @@ string GetBranch()
...
+            {
+                (errors, config) = YamlUtility.Deserialize<JObject>(File.ReadAllText(filePath));
+            }
+            else if (filePath.Contains(".json", StringComparison.OrdinalIgnoreCase))
+            {
+                (errors, config) = JsonUtility.Deserialize<JObject>(File.ReadAllText(filePath));
```
Comment (yufeih) [inline src/docfx/config/Config.cs]: Some of the config error file are hard coding `docfx.yml`

File: src/docfx/config/Config.cs (hunk 1 above)
Comment (yufeih) [reply src/docfx/config/Config.cs]: interesting, maybe we should be using the name before calling GetRestorePath?

File: src/docfx/docset/Docset.cs [hunk 3]
```diff
@@ -59,6 +59,11 @@ internal class Docset
         /// </summary>
         public HashSet<Document> BuildScope => _buildScope.Value;
 
+        /// <summary>
+        /// Gets the config file name.
+        /// </summary>
+        public string ConfigFile { get; }
```
Comment (OsmondJiang) [inline src/docfx/docset/Docset.cs]: >ConfigFile [](start = 22, length = 10)

ConfigFileName?

Comment (yufeih) [reply src/docfx/docset/Docset.cs]: Other places uses `XXXPath`, so `ConfigPath`?

Comment (OsmondJiang) [reply src/docfx/docset/Docset.cs]: ConfigPath like a relative path, but it;s OK

---
In reply to: [225480012](https://github.com/dotnet/docfx/pull/3498#discussion_r225480012) [](ancestors = 225480012)

File: src/docfx/config/Config.cs [hunk 4]
```diff
@@ -152,14 +152,14 @@ public IEnumerable<string> GetExternalReferences()
         /// <summary>
         /// Load the config under <paramref name="docsetPath"/>
         /// </summary>
-        public static (List<Error>, Config) Load(string docsetPath, CommandLineOptions options, bool extend = true, RestoreMap restoreMap = null)
+        public static (List<Error>, Config, string) Load(string docsetPath, CommandLineOptions options, bool extend = true, RestoreMap restoreMap = null)
```
Comment (OsmondJiang) [inline src/docfx/config/Config.cs]: >List<Error>, Config, string [](start = 23, length = 27)

(List<Error> errors, Config config, string configPath/ConfigName)?

File: docs/specs/config.yml [hunk 5]
```diff
@@ -1,10 +1,14 @@
...
+# Specify include files, prefer docfx.yml to docfx.json
 inputs:
   docfx.yml: |
     content:
       include:
         - docs/**/*.md
+  docfx.json: |
+    content:
+      include:
+        - ignore.md
```
Comment (OsmondJiang) [inline docs/specs/config.yml]: why json file is in yml formate ...

File: src/docfx/docset/Docset.cs (hunk 3 above)
Comment (live1206) [reply src/docfx/docset/Docset.cs]: ConfigPath is something else, would take ConfigFilaName here as below:
```csharp
public static (List<Error> errors, Config config, string configFileName) Load(string docsetPath, CommandLineOptions options, bool extend = true, RestoreMap restoreMap = null)
```

File: src/docfx/docset/Docset.cs [hunk 6]
```diff
@@ -67,11 +72,12 @@ internal class Docset
         private readonly Lazy<RedirectionMap> _redirections;
         private readonly Lazy<LegacyTemplate> _legacyTemplate;
 
-        public Docset(Context context, string docsetPath, Config config, CommandLineOptions options)
+        public Docset(Context context, string docsetPath, Config config, CommandLineOptions options, string configFile)
```
Comment (OsmondJiang) [inline src/docfx/docset/Docset.cs]: >Config config, CommandLineOptions options, string configFile [](start = 58, length = 60)

that seems weired when the parameters have both config and its name, if the configurartion is always load from Config.Load, why config doesn't know it's name?

**Pull request merged**